syncer/db.go

```diff
@@ -69,45 +69,88 @@ type binlogSize struct {
-		return terror.ErrDBUnExpect.Generate("database base connection not valid")
-	}
-	return conn.baseConn.ResetConn(tctx)
+// UpStreamConn connect to upstream DB, use *sql.DB instead of *sql.Conn
+type UpStreamConn struct {
+	cfg    *config.SubTaskConfig
+	BaseDB *conn.BaseDB
+}
+
+func (conn *UpStreamConn) getMasterStatus(flavor string) (mysql.Position, gtid.Set, error) {
+	return utils.GetMasterStatus(conn.BaseDB.DB, flavor)
+}
+
+func (conn *UpStreamConn) getServerUUID(flavor string) (string, error) {
+	return utils.GetServerUUID(conn.BaseDB.DB, flavor)
+}
+
+func (conn *UpStreamConn) getParser(ansiQuotesMode bool) (*parser.Parser, error) {
+	return utils.GetParser(conn.BaseDB.DB, ansiQuotesMode)
 }
 
-func (conn *Conn) getMasterStatus(flavor string) (mysql.Position, gtid.Set, error) {
-	return utils.GetMasterStatus(conn.baseConn.DB, flavor)
+func (conn *UpStreamConn) killConn(connID uint32) error {
+	return utils.KillConn(conn.BaseDB.DB, connID)
 }
 
-func (conn *Conn) getServerUUID(flavor string) (string, error) {
-	return utils.GetServerUUID(conn.baseConn.DB, flavor)
+func (conn *UpStreamConn) fetchAllDoTables(bw *filter.Filter) (map[string][]string, error) {
+	return utils.FetchAllDoTables(conn.BaseDB.DB, bw)
 }
 
-func (conn *Conn) getParser(ansiQuotesMode bool) (*parser.Parser, error) {
-	return utils.GetParser(conn.baseConn.DB, ansiQuotesMode)
+func (conn *UpStreamConn) countBinaryLogsSize(pos mysql.Position) (int64, error) {
+	return countBinaryLogsSize(pos, conn.BaseDB.DB)
 }
 
-func (conn *Conn) killConn(connID uint32) error {
-	return utils.KillConn(conn.baseConn.DB, connID)
+func createUpStreamConn(cfg *config.SubTaskConfig, dbCfg config.DBConfig) (*UpStreamConn, error) {
+	baseDB, err := createBaseDB(dbCfg)
+	if err != nil {
+		return nil, terror.WithScope(terror.DBErrorAdapt(err, terror.ErrDBDriverError), terror.ScopeUpstream)
+	}
+	return &UpStreamConn{BaseDB: baseDB, cfg: cfg}, nil
 }
 
-func (conn *Conn) fetchAllDoTables(bw *filter.Filter) (map[string][]string, error) {
-	return utils.FetchAllDoTables(conn.baseConn.DB, bw)
+func closeUpstreamConn(tctx *tcontext.Context, conn *UpStreamConn) {
+	err := conn.BaseDB.Close()
+	if err != nil {
+		tctx.L().Error("fail to close baseConn connection", log.ShortError(err))
+	}
 }
 
-func (conn *Conn) countBinaryLogsSize(pos mysql.Position) (int64, error) {
-	return countBinaryLogsSize(pos, conn.baseConn.DB)
+// WorkerConn represents a live DB connection
+type WorkerConn struct {
+	cfg      *config.SubTaskConfig
+	baseConn *conn.BaseConn
+
+	// for workerConn to reset baseConn
+	resetConnFn func(*tcontext.Context) (*conn.BaseConn, error)
+}
+
+// ResetConn reset one worker connection from specify *BaseDB
+func (conn *WorkerConn) resetConn(tctx *tcontext.Context) error {
+	if conn == nil {
+		return terror.ErrDBDriverError.Generate("database not valid")
+	}
+	dbConn, err := conn.resetConnFn(tctx)
+	if err != nil {
+		return err
+	}
+	if conn.baseConn != nil {
+		err = conn.baseConn.Close()
```

Comment:
Is it better to put the connection `Close` logic into `resetConnFn`?

Reply:
If we don't close connection, it's closed by golang GC mechanism, it looks better.
I'll fix it